### ISSUE TITLE
JSX elements not syntax highlighted correctly

### ISSUE THREAD

**Issue by trygveaa** [neoclide/vim-jsx-improve#6]
I'm having some issues with syntax highlighting of JSX with this plugin. Initially when I open a file the JSX elements and attributes are white. However after using vim for a while (changing files, writing code, etc.), the highlighting suddenly worked. I'm not sure how that happened and can't reproduce it consistently. If I close vim and open the file again, the highlighting is broken again.

A screenshot of two vim instances. On the left is a newly opened vim where highlighting is broken. On the right is one where it has started working (I pasted in the code from the `react.js` file in this repo).

![image](https://cloud.githubusercontent.com/assets/601966/25194745/b5a819ee-253b-11e7-803b-2015220152d7.png)

Do you know what could be wrong, or how I can debug it?

**Comment by carleryd** [comment]
I have the same problem, and it seem to trigger the colors whenever I enter a .html file and then back to .js in the same window!

**Comment by emarcotte** [comment]
have been able to get it to work with: 
```
set ft=html
set ft=javascript
```
as a quick hack when i notice the file doesn't look right. Not sure how to automatically do it yet.

**Comment by rosenfeld** [comment]
@emarcotte as a workaround until this is fixed you might want to add this work-around in your vim initializers:

```vim
au BufNewFile,BufRead *.jsx setlocal ft=html ft=javascript
```

This will force vim to load the html syntax highlighter whenever a jsx is loaded. Maybe there's a better way to only do that in the first load of some jsx, or to force Vim loading the html support upon initialization. If you load any html file before jsx, this hack is not required by the way. It seems we just need to load html support at some point...

**Comment by chemzqm** [comment]
@trygveaa  I'm sorry for didn't responsive for some while, I think the problem is conflict with other syntax plugin like `vim-javascript`,  please try to disable other vim plugins and check if the problem still exists.

It should works without `setlocal ft=html`

**Comment by bwells** [comment]
I also have this issue. I don't have any other javascript syntax plugins installed.

My list of javascript syntax files (derived from `:let @a=globpath(&rtp, 'syntax/javascript.vim')` is

~/.vim/plugged/vim-jsx-improve/syntax/javascript.vim
/usr/local/share/vim/vim80/syntax/javascript.vim
~/.vim/plugged/vim-jsx-improve/after/syntax/javascript.vim

**Comment by chemzqm** [comment]
@bwells  could be caused by not have `filetype plugin indent on` in your `.vimrc`.
Try add `autocmd BufEnter *.js set ft=javascript` to your `.vimrc` and reopen js file to see if it works.

**Comment by bwells** [comment]
This does not appear to be the issue. It's enabled by default via vim-sensible and checking `:filetype` yields `filetype detection:ON  plugin:ON  indent:ON`.

I've tried disabling all plugins except for vim-sensible and vim-jsx-improve with the same results. 

Explicitly setting the filetype to javascript, even when it's already auto detected correctly, does not help with the syntax highlighting. The `filetype=html filetype=javascript` trick does work however. 

I was able to reproduce that opening a file with the `html` filetype BEFORE ever opening a file with the `javascript` filetype ended up with the correct syntax highlighting. This negated the need for the `ft=html ft=javascript` trick. 

It seems like loading the html filetype into memory is required before JSX syntax highlighting will correctly and this is why the `ft=html ft=javascript` trick works. I tried with xml files before javascript and that did not work.
